### PR TITLE
[RLlib] Performance booster (cache in- and out-) specs RLModule; MetricsLogger speedups, avoid NestedDict, etc..

### PR DESCRIPTION
**Pull request by sven1977**
Performance booster:
* Cache all RLModule checking (in- and output);
* MetricsLogger speedups;
* avoid NestedDict;
* Only create OneHotCategorlcal when really needed.

Speedup measured:

```
from ray.rllib.algorithms.impala import ImpalaConfig
from ray.rllib.env.single_agent_env_runner import SingleAgentEnvRunner
from timeit import timeit

config = (
    ImpalaConfig()
    .api_stack(
        enable_rl_module_and_learner=True,
        enable_env_runner_and_connector_v2=True,
    )
    .env_runners(create_env_on_local_worker=True)
    .environment("CartPole-v1")
    .rl_module(
        model_config_dict={
            "vf_share_layers": True,
            "uses_new_env_runners": True,
        }
    )
)

print(timeit(
    stmt="env_runner.sample(); env_runner.get_metrics()",
    setup="env_runner = SingleAgentEnvRunner(config=config)",
    number=500,
    globals={"config": config, "SingleAgentEnvRunner": SingleAgentEnvRunner},
))
```

With this PR:
6.4 sec

W/o this PR:
10.7 sec

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/models/specs/checker.py [hunk 1]
```diff
@@ -173,7 +173,7 @@ def check_input_specs(
     *,
     only_check_on_retry: bool = True,
     filter: bool = False,
-    cache: bool = False,
+    cache: bool = True,
```
Comment (simonsays1980) [inline rllib/core/models/specs/checker.py:176]: Wow, does that mean we never cached the checks in the RLModules?

File: rllib/env/multi_agent_env_runner.py [hunk 2]
```diff
@@ -813,29 +813,34 @@ def _make_on_episode_callback(self, which: str, episode=None):
...
-            {
-                NUM_ENV_STEPS_SAMPLED_LIFETIME: num_steps,
-                # TODO (sven): obs is not-vectorized. Support vectorized MA envs.
-                NUM_AGENT_STEPS_SAMPLED_LIFETIME: {str(aid): 1 for aid in next_obs},
-                NUM_MODULE_STEPS_SAMPLED_LIFETIME: {
-                    episode.module_for(aid): 1 for aid in next_obs
-                },
-            },
-            reduce="sum",
+        self.metrics.log_value(
```
Comment (simonsays1980) [inline rllib/env/multi_agent_env_runner.py:816]: Shouldn't we use `log_dict` anymore?

File: rllib/env/single_agent_env_runner.py [hunk 3]
```diff
@@ -556,6 +558,8 @@ def _sample_episodes(
         # Initialized episodes have to be removed as they lack `extra_model_outputs`.
         samples = [episode for episode in done_episodes_to_return if episode.t > 0]
 
+        self._increase_sampled_metrics(ts)
```
Comment (simonsays1980) [inline rllib/env/single_agent_env_runner.py:561]: Nice, much faster.

File: rllib/env/single_agent_env_runner.py [hunk 4]
```diff
@@ -739,56 +743,54 @@ def _convert_to_tensor(self, struct) -> TensorType:
...
-        self.metrics.log_dict(
-            {
-                "episode_len_min": length,
-                "episode_return_min": ret,
-            },
-            reduce="min",
+        # To mimick the old API stack behavior, we'll use `window` here for
+        # these particular stats (instead of the default EMA).
+        win = self.config.metrics_num_episodes_for_smoothing
+        self.metrics.log_value("episode_len_mean", length, window=win)
```
Comment (simonsays1980) [inline rllib/env/single_agent_env_runner.py:780]: Can we define constants for `episode_len_mean`, etc. We use them a lot.

File: rllib/models/torch/torch_distributions.py [hunk 5]
```diff
@@ -113,11 +113,12 @@ def __init__(
 
         self.probs = probs
         self.logits = logits
-        self.one_hot = torch.distributions.one_hot_categorical.OneHotCategorical(
-            logits=logits, probs=probs
-        )
         super().__init__(logits=logits, probs=probs)
 
+        # Build this distribution only if really needed (in `self.rsample()`). It's
```
Comment (simonsays1980) [inline rllib/models/torch/torch_distributions.py:118]: Sweet. So Sampling increases in performance even more.

**Review by simonsays1980 — APPROVED** [review]
LGTM: Excited to see what's the improvement in performance.

File: rllib/env/multi_agent_env_runner.py (hunk 2 above)
Comment (sven1977) [reply rllib/env/multi_agent_env_runner.py:816]: We can still use it, but it's slower, b/c it has to come up with the nested key-sequences itself, whereas they are provided by the user when using `log_value`. I have to think about MetricsLogger performance a bit more. There are some caveats when using it excessively.

File: rllib/core/models/specs/checker.py (hunk 1 above)
Comment (sven1977) [reply rllib/core/models/specs/checker.py:176]: Never - by default - for the outputs, correct.

**Pull request merged**